### ISSUE TITLE
ng serve fails with "ERROR in Debug failure: no local name provided for NamespaceImport"

### ISSUE THREAD

**Issue by alcfeoh** [angular/angular#36006]
# 🐞 Bug report

### Command (mark with an `x`)

- [ ] new
- [ ] build
- [x] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

### Is this a regression?

Yes, the same app was working with Angular 8 and CLI 8.

### Description

 A clear and concise description of the problem...
Running ng serve consistently fails with the stack trace reported below.

## 🔬 Minimal Reproduction
npm install
ng serve
It's a private repo and the exception does not refer to any specific code in the repo. It seems to be a generic error in compiler-cli itself.

## 🔥 Exception or Error
<pre><code>
ERROR in Debug failure: no local name provided for NamespaceImport
F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\core\src\compiler.js:342
                finally { if (e_2) throw e_2.error; }
                                   ^

Error: Debug failure: no local name provided for NamespaceImport
    at extractModuleAndNameFromImport (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\type_to_value.js:173:27)
    at Object.typeToValue (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\type_to_value.js:59:28)
    at F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\typescript.js:76:58
    at Array.map (<anonymous>)
    at TypeScriptReflectionHost.getConstructorParameters (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\typescript.js:55:36)
    at Object.getConstructorDependencies (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\util.js:32:36)
    at extractInjectableCtorDeps (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\injectable.js:233:38)
    at InjectableDecoratorHandler.analyze (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\injectable.js:71:31)
    at TraitCompiler.analyzeTrait (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\compilation.js:341:40)
    at analyze (F:\webstr\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\compilation.js:293:58)

</code></pre>

## 🌍 Your Environment
<pre><code>
Angular CLI: 9.0.5
Node: 12.13.1
OS: win32 x64

Angular: 9.0.5
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.5
@angular-devkit/build-angular     0.900.5
@angular-devkit/build-optimizer   0.900.5
@angular-devkit/build-webpack     0.900.5
@angular-devkit/core              9.0.5
@angular-devkit/schematics        9.0.5
@angular/cdk                      9.1.2
@angular/material                 9.1.2
@ngtools/webpack                  9.0.5
@schematics/angular               9.0.5
@schematics/update                0.900.5
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2

</code></pre>

**Anything else relevant?**

**Comment by avoliva** [comment]
Any update?

**Comment by alcfeoh** [comment]
@avoliva Do you have a similar issue?

**Comment by avoliva** [comment]
Yes, I think I just figured it out though. Here's a link that helped: https://www.gitmemory.com/issue/rollbar/rollbar.js/806/561361290

Basically, new typescript doesn't support the `import * as x from Library` syntax. You have to change it to `import x from Library` and then in your tsconfig file add this to the compilerOptions `"allowSyntheticDefaultImports": true`

**Comment by JoostK** [comment]
The statement that TS no longer supports `import * as x from 'y';` is not true, that's perfectly valid. Tracing back the error indicates that there's likely a nested reference to a type originating from the namespace import:

```ts
import * as x from 'x';

@Injectable()
class MyService {
  constructor(private foo: x.Foo.Bar) {}
}
```

The relevant part is the `x.Foo.Bar` type reference in the constructor's parameter type. Such a reference is not currently supported (only one level, such as `x.Foo`, currently works).

Could you check whether such a type reference is used somewhere?

**Comment by avoliva** [comment]
Not in my code. However, I was using the library that link I posted was referencing (Rollbar) so it's possible there was a reference like that somewhere in there. However, when I made those changes I talked about earlier any errors went away.

**Comment by JoostK** [comment]
@avoliva Can you share the full diff that was required to make it work? If you are able to change an import then I don't see how the library could be broken.

**Comment by avoliva** [comment]
The only change I had to make was making my import of Rollbar from `import * as Rollbar from 'rollbar'` to `import Rollbar from 'rollbar'`

After that, I added `"allowSyntheticDefaultImports": true` to the compiler options in my tsconfig file.

**Comment by JoostK** [comment]
@avoliva Oh, I see. How are you using that import, then? I presume it's used in a position where Angular needs to do something with it.

**Comment by avoliva** [comment]
Yeah I have a rollbar.ts file that looks like this:

    import {
      Injectable,
      Inject,
      InjectionToken,
      ErrorHandler
    } from '@angular/core';

    import Rollbar from 'rollbar';


    const rollbarConfig = {
      accessToken: 'XXX',
      captureUncaught: true,
      captureUnhandledRejections: true,
      cookies: document.cookie,
    };

    export const RollbarService = new InjectionToken<Rollbar>('rollbar');

    @Injectable()
    export class RollbarErrorHandler implements ErrorHandler {
      constructor(
        @Inject(RollbarService) private rollbar: Rollbar) {
      }

      handleError(err: any): void {
        let error = err.originalError || err;
        if (window.location.hostname !== "localhost") {
          this.rollbar.error(error);
        }
      }
    }

    export function rollbarFactory() {
      return new Rollbar(rollbarConfig);
    }

Then in my app.component.ts file I import both the RollbarService and Rollbar

    import { RollbarService } from './lib/rollbar';
    import * as Rollbar from 'rollbar';

In the constructor for the AppComponent, I add this:

    @Inject(RollbarService) private rollbar: Rollbar,

That's the whole implementation

**Comment by alcfeoh** [comment]
> The only change I had to make was making my import of Rollbar from `import * as Rollbar from 'rollbar'` to `import Rollbar from 'rollbar'`
> 
> After that, I added `"allowSyntheticDefaultImports": true` to the compiler options in my tsconfig file.

@avoliva This was the exact same issue for me. My project is also using rollbar. Thanks for the tip, the code does compile now.

Now, from an Angular 9 perspective, it still looks like a regression because that code worked with Angular 8.

**Comment by JoostK** [comment]
@avoliva Ah, excellent thanks for the details. The default import seems like the proper workaround here. The namespace import like `import * as Rollbar from 'rollbar';` cannot be used as a value (which Angular's DI system requires) as TypeScript would then complain:

> A namespace-style import cannot be called or constructed, and will cause a failure at runtime. Consider using a default import or import require here instead

In this particular case however, the usage of `@Inject` makes the type-to-value conversion irrelevant as an explicit token is provided, so it looks like we should not be throwing that error. I experimented with that change in the compiler and this would properly produce a diagnostic if `@Inject` is omitted and `strictInjectionParameters` is enabled:

```
error NG2003: No suitable injection token for parameter 'rollbar' of class 'RollbarErrorHandler'.
Found Rollbar

  constructor(
    @Inject(RollbarService) private rollbar: Rollbar) {
                                    ~~~~~~~

  }
```

**Comment by JoostK** [comment]
Tracking as FW-1995 (internal only)

**Comment by Maximaximum** [comment]
I'm having the same issue, even though I'm not using Rollbar.

```
Angular CLI: 9.0.6
Node: 12.13.1
OS: win32 x64

Angular: 9.0.6
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, localize, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.6
@angular-devkit/build-angular     0.900.6
@angular-devkit/build-optimizer   0.900.6
@angular-devkit/build-webpack     0.900.6
@angular-devkit/core              9.0.6
@angular-devkit/schematics        9.0.6
@ngtools/webpack                  9.0.6
@schematics/angular               9.0.6
@schematics/update                0.900.6
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
```

```
ng serve

chunk {main} main.js, main.js.map (main) 1.99 kB [initial] [rendered]
chunk {polyfills} polyfills.js, polyfills.js.map (polyfills) 670 bytes [initial] [rendered]
chunk {runtime} runtime.js, runtime.js.map (runtime) 6.15 kB [entry] [rendered]
chunk {styles} styles.css, styles.css.map (styles) 467 kB [initial] [rendered]
chunk {vendor} vendor.js, vendor.js.map (vendor) 339 kB [initial] [rendered]
Date: 2020-03-16T16:52:45.227Z - Hash: a9e5f886e61b5db022fc - Time: 9722ms

WARNING in Conflict: Multiple assets emit different content to the same filename loading.gif

ERROR in Debug failure: no local name provided for NamespaceImport
** Angular Live Development Server is listening on localhost:4200, open your browser on https://localhost:4200/ **
```

**Comment by Maximaximum** [comment]
And strangely enough, if I wait for about 30 minutes, the following line gets also emitted:

```

C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\core\src\compiler.js:342
                finally { if (e_2) throw e_2.error; }
                                   ^

Error: Debug failure: no local name provided for NamespaceImport
    at extractModuleAndNameFromImport (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\type_to_value.js:173:27)
    at Object.typeToValue (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\type_to_value.js:59:28)
    at C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\typescript.js:76:58
    at Array.map (<anonymous>)
    at TypeScriptReflectionHost.getConstructorParameters (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\reflection\src\typescript.js:55:36)
    at Object.getConstructorDependencies (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\util.js:32:36)
    at extractInjectableCtorDeps (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\injectable.js:233:38)
    at InjectableDecoratorHandler.analyze (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\annotations\src\injectable.js:71:31)
    at TraitCompiler.analyzeTrait (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\compilation.js:341:40)
    at analyze (C:\Projects\NGP\Frontend\node_modules\@angular\compiler-cli\src\ngtsc\transform\src\compilation.js:293:58)

Date: 2020-03-16T17:09:35.609Z - Hash: 8dde01b507526d871075
5 unchanged chunks

Time: 861ms

WARNING in AngularCompilerPlugin: Forked Type Checker exited unexpectedly. Falling back to type checking on main thread.

WARNING in Conflict: Multiple assets emit different content to the same filename loading.gif

ERROR in Debug failure: no local name provided for NamespaceImport
```

**Comment by JoostK** [comment]
@Maximaximum  Could you change this line as follows:

https://github.com/angular/angular/blob/ccb70e1c641d38c410d0392c9071e5ee309ddc89/packages/compiler-cli/src/ngtsc/reflection/src/type_to_value.ts#L174

```diff
-        throw new Error(`Debug failure: no local name provided for NamespaceImport`);
+        throw new Error(`Debug failure: no local name provided for NamespaceImport: ${node.getSourceFile().fileName}`);
```

And then share the details of that file (the import statements and the constructor)

**Comment by Maximaximum** [comment]
@JoostK Thank you. This is what that file looks like:

```typescript
import { Inject, Injectable, InjectionToken } from '@angular/core';
import * as MemoryStorage from 'memorystorage';

import { StorageWrapper } from '../storage';

export const MEMORY_STORAGE = new InjectionToken<MemoryStorage>('MEMORY_STORAGE', {
    providedIn: 'root',
    factory: () => new MemoryStorage()
});

@Injectable({
    providedIn: 'root'
})
export class MemoryStorageService extends StorageWrapper {
    constructor(@Inject(MEMORY_STORAGE) memoryStorage: MemoryStorage) {
        super(memoryStorage);
    }
}
```

So, the error is caused by the line `import * as MemoryStorage from 'memorystorage';`
When I convert it to `import MemoryStorage from 'memorystorage';` and use `    "allowSyntheticDefaultImports": true`, the error is gone.

I wonder whether this is just a bug in Angular, or this is the expected behavior?

**Comment by JoostK** [comment]
@Maximaximum This is definitely a bug in the compiler. The compiler attempts to convert the `MemoryStorage` type declaration (for the `memoryStorage` parameter in the constructor) to a value expression that is used as DI token. In this case, it doesn't know how to handle the namespace import (it's actually invalid to refer to `MemoryStorage` as a value, the workaround using `"allowSyntheticDefaultImports"` is then required to express `MemoryStorage` as a value), so it throws.

However, the type-to-value conversion is irrelevant in this case because there an explicit DI token provided using the `@Inject` decorator. As such, the compiler should not crash when it fails to convert a type to a value, as that may be handled using an `@Inject` decorator.

**Comment by Maximaximum** [comment]
@JoostK could you please give a hint if this fix will soon appear in an Angular v9 patch release?

**Comment by JoostK** [comment]
@Maximaximum It's been merged into the 9.1.x branch just after the release of 9.1.1, so it will be included in 9.1.2 next week. Releases typically go out on Wednesdays.
